File: src/Core/NameDatabase.cs
// This database provides a rich variety of names and concepts that blend historical, contemporary, and futuristic elements, with a focus on solarpunk and cyberpunk themes. Here's a breakdown of the categories:
// 1. NPC Names: A mix of modern, futuristic, and nature-inspired names.
// 2. Node Names: Locations that combine technological and ecological concepts.
// 3. Corporation Names: Companies that focus on sustainable tech and futuristic innovations.
// 4. Group Names: Organizations and movements that reflect various subcultures and ideologies.
// 5. Culture Names: Philosophical and cultural movements that might emerge in a solarpunk future.
// You can easily expand this database by adding more entries to each category or creating new categories as needed. This data can be loaded into your game to randomly generate names for NPCs, locations, and other entities, helping to create a rich and diverse game world.
// To use this in your C# code, you could create a class to load and manage this data:


using System.IO;
using Newtonsoft.Json;

public class NameDatabase
{
    public NameData Data { get; private set; }

    public NameDatabase(string filePath)
    {
        string json = File.ReadAllText(filePath);
        Data = JsonConvert.DeserializeObject<NameData>(json);
    }

    public string GetRandomNPCName()
    {
        string firstName = GetRandomElement(Data.npc_names.first_names);
        string lastName = GetRandomElement(Data.npc_names.last_names);
        return $"{firstName} {lastName}";
    }

    public string GetRandomNodeName()
    {
        return GetRandomElement(Data.node_names);
    }

    // Add similar methods for other categories

    private string GetRandomElement(string[] array)
    {
        return array[new Random().Next(array.Length)];
    }
}

public class NameData
{
    public NPCNames npc_names { get; set; }
    public string[] node_names { get; set; }
    public string[] corporation_names { get; set; }
    public string[] group_names { get; set; }
    public string[] culture_names { get; set; }
}

public class NPCNames
{
    public string[] first_names { get; set; }
    public string[] last_names { get; set; }
}

File: src/Core/NameDatabase.cpp
#include "NameDatabase.hpp"
#include <godot_cpp/classes/file_access.hpp>

void NameDatabase::_bind_methods() {
    ClassDB::bind_method(D_METHOD("load_from_file", "file_path"), &NameDatabase::load_from_file);
    ClassDB::bind_method(D_METHOD("get_random_npc_name"), &NameDatabase::get_random_npc_name);
    ClassDB::bind_method(D_METHOD("get_random_node_name"), &NameDatabase::get_random_node_name);
    ClassDB::bind_method(D_METHOD("get_random_corporation_name"), &NameDatabase::get_random_corporation_name);
    ClassDB::bind_method(D_METHOD("get_random_group_name"), &NameDatabase::get_random_group_name);
    ClassDB::bind_method(D_METHOD("get_random_culture_name"), &NameDatabase::get_random_culture_name);
}

NameDatabase::NameDatabase() : gen(rd()) {}

void NameDatabase::load_from_file(const String& file_path) {
    Ref<FileAccess> f = FileAccess::open(file_path, FileAccess::READ);
    if (f.is_valid()) {
        String json_str = f->get_as_text();
        Variant parsed = JSON::parse_string(json_str);
        
        if (parsed.get_type() == Variant::DICTIONARY) {
            Dictionary data = parsed;
            
            Dictionary npc_names = data.get("npc_names", Dictionary());
            npc_first_names = npc_names.get("first_names", Array());
            npc_last_names = npc_names.get("last_names", Array());
            
            node_names = data.get("node_names", Array());
            corporation_names = data.get("corporation_names", Array());
            group_names = data.get("group_names", Array());
            culture_names = data.get("culture_names", Array());
        }
    }
}

String NameDatabase::get_random_npc_name() const {
    String first_name = get_random_element(npc_first_names);
    String last_name = get_random_element(npc_last_names);
    return first_name + " " + last_name;
}

String NameDatabase::get_random_node_name() const {
    return get_random_element(node_names);
}

String NameDatabase::get_random_corporation_name() const {
    return get_random_element(corporation_names);
}

String NameDatabase::get_random_group_name() const {
    return get_random_element(group_names);
}

String NameDatabase::get_random_culture_name() const {
    return get_random_element(culture_names);
}

String NameDatabase::get_random_element(const Array& arr) const {
    if (arr.size() == 0) {
        return "";
    }
    std::uniform_int_distribution<> dis(0, arr.size() - 1);
    return arr[dis(gen)];
} 